Lesson 04/news_reader/index.js

```diff
@@ -0,0 +1,269 @@
+const app = require('express')();
+const bodyParser = require('body-parser');
+const urlencodedParser = bodyParser.urlencoded({extended: false});
+const consolidate = require('consolidate');
+const request = require('request');
+const path = require('path');
+const fs = require('fs');
+const cheerio = require('cheerio');
+
+app.cli = require('cli');
+app.cli = require('cli');
+app.DEFAULT_PORT = 3000;
+
+// Set command line options.
+app.cli.parse({
+	serverPort: ['p', 'Bind a custom port for requests (1024-65535).', 'int', app.DEFAULT_PORT]
+});
+
+if (app.cli.options.serverPort < 1024 || app.cli.options.serverPort > 65535) {
+	console.warning(
+		`Invalid custom port defined (${app.cli.options.serverPort}). ` +
+		`Fallback to the default port ${app.DEFAULT_PORT}.`
+	);
+	
+	app.cli.options.serverPort = app.DEFAULT_PORT;
+}
+
+// Application settings for internal use.
+app.viewSettings = {
+	pageTitle: `NewsFlash | A simple news fetch service`,
+	mainPage: `http://localhost:${app.cli.options.serverPort}`,
+	newsPage: `http://localhost:${app.cli.options.serverPort}/news`
+};
+
+app.siteSettings = {
+	vesti: {
+		name: 'vesti',
+		title: 'vesti.ru',
+		articleUrl: 'https://www.vesti.ru/',
+		requestUrl: 'https://www.vesti.ru/news/',
+		categories: []
+	}
+};
+
+
+app.engine('hbs', consolidate.handlebars);
+app.set('view engine', 'hbs');
+app.set('views', path.join(__dirname, 'views'));
+
+
+// News category parser functions.
+function parseCategoryVesti($) {
+	let cid;
+	
+	$('.aside-menu__list-item', '.aside-menu__list').each((i, $elem) => {
+		$elem = $('a', $elem);
+		cid = $elem.attr('href').trim().match(/cid=(\d+)$/);
+		
+		if (cid) {
+			app.siteSettings.vesti.categories.push({
+				cid: cid[1],
+				title: $elem.text()
+			});
+		}
+	});
+	
+	return (app.siteSettings.vesti.categories.length ? app.siteSettings.vesti.categories : null);
+}
+
+const responseData = {
+	result: 0,
+	message: '',
+	data: null
+};
+
+
+function fetchNewsVesti(clientReq, clientRes) {
+	let requestUrl = app.siteSettings.vesti.requestUrl;
+	
+	if (clientReq.body.newsCategory) {
+		requestUrl =
+			`${app.siteSettings.vesti.articleUrl}section.html?` +
+			`cid=${clientReq.body.newsCategory}`;
+	}
+	
+	request(requestUrl, (err, res, body) => {
+		const news = [];
+		const newsCount = parseInt(clientReq.body.newsCount);
+		let category;
+
+		if (!err && res.statusCode === 200) {
+			const $ = cheerio.load(body);
+			let $newsWrapper = '.news-wrapper_';
+			
+			if (clientReq.body.newsCategory) {
+				$newsWrapper = '.news-wrapper_section';
+				
+				app.siteSettings.vesti.categories.forEach((elem) => {
+					if (elem.cid === clientReq.body.newsCategory) {
+						category = elem.title;
+					}
+				});
+			}
+			
+			$('.b-item__inner', $newsWrapper).each((i, $elem) => {
+				// Skip exeeded amount of news items.
+				if (newsCount && i >= newsCount) {
+					return;
+				}
+				
+				const $title = $('.b-item__title a', $elem);
+
+				news.push({
```

Comment:
The preferable way - don't mutate objects/arrays

Reply:
ACK.
Will try to redo work logic with this object.